ml-agents/mlagents/trainers/subprocess_env_manager.py

```diff
@@ -306,10 +315,31 @@ def training_behaviors(self) -> Dict[BehaviorName, BehaviorSpec]:
 
     def close(self) -> None:
         logger.debug("SubprocessEnvManager closing.")
+        for env_worker in self.env_workers:
+            env_worker.request_close()
+        # Pull messages out of the queue until every worker has CLOSED or we time out.
+        deadline = datetime.now() + WORKER_SHUTDOWN_TIMEOUT
+        while self.workers_alive > 0 and datetime.now() < deadline:
+            try:
+                step: EnvironmentResponse = self.step_queue.get_nowait()
+                env_worker = self.env_workers[step.worker_id]
+                if step.cmd == EnvironmentCommand.CLOSED and not env_worker.closed:
+                    env_worker.closed = True
+                    self.workers_alive -= 1
+                # Discard all other messages.
+            except EmptyQueueException:
+                pass
         self.step_queue.close()
+        # Sanity check to kill zombie workers and report an issue if they occur.
+        if self.workers_alive > 0:
+            logger.error("SubprocessEnvManager had workers that didn't signal shutdown")
+            for env_worker in self.env_workers:
+                if not env_worker.closed and env_worker.process.is_alive():
+                    env_worker.process.terminate()
+                    raise AssertionError(
+                        "A SubprocessEnvManager worker did not shut down correctly"
+                    )
```

Comment:
I am not sure if we need this, it may be good enough to log the warning.

Reply:
I think logging is better here; raising will mean that we don't try to shut down other workers that might be stuck, and we'd have to handle the error higher up too.